src/cargo/core/compiler/fingerprint/mod.rs

```diff
@@ -796,14 +798,36 @@ pub enum StaleItem {
 impl LocalFingerprint {
     /// Read the environment variable of the given env `key`, and creates a new
     /// [`LocalFingerprint::RerunIfEnvChanged`] for it.
-    ///
-    // TODO: This is allowed at this moment. Should figure out if it makes
-    // sense if permitting to read env from the config system.
     #[allow(clippy::disallowed_methods)]
-    fn from_env<K: AsRef<str>>(key: K) -> LocalFingerprint {
+    fn from_env<K: AsRef<str>>(
+        key: K,
+        env_config: &Arc<HashMap<String, OsString>>,
+        env_config_insensitive: &Arc<LazyCell<HashMap<String, OsString>>>,
+    ) -> LocalFingerprint {
         let key = key.as_ref();
         let var = key.to_owned();
-        let val = env::var(key).ok();
+        let val = if let Some(val) = env::var(key).ok() {
+            Some(val)
+        } else {
+            match env_config.get(key) {
```

Comment:
which is supposed to have higher precedence?

Reply:
I think `env_config` should have higer precedence as `env_config`  may contains the env that needs  force overwrite or that doesn't exist in the env snapshot